stackdriver/config/config.go

```diff
@@ -0,0 +1,168 @@
+	// Required when the Kubernetes cluster is not hosted on GCE.
+	GCPLocation string
+	// ClusterName is the cluster name with which the data will be associated in Stackdriver.
+	// Required when the Kubernetes cluster is not hosted on GCE.
+	ClusterName string
+	// GCPSecretName is the optional GCP service account key which will be used to
+	// authenticate with Stackdriver. If not provided, Google Application Default Credentials
+	// will be used (https://cloud.google.com/docs/authentication/production).
+	GCPSecretName string
+	// GCPSecretNamespace is the Kubernetes namespace where GcpSecretName is located.
+	// The Kubernetes ServiceAccount used by the pod that is exporting data to
+	// Stackdriver should have access to Secrets in this namespace.
+	GCPSecretNamespace string
+}
+
+// NewStackdriverConfigFromConfigMap returns a Config for the given configmap
+func NewStackdriverConfigFromConfigMap(config *corev1.ConfigMap) (*Config, error) {
+	if config == nil {
+		return &Config{}, nil
+	}
+	return NewStackdriverConfigFromMap(config.Data)
+}
+
+// NewStackdriverConfigFromMap returns a Config for the given map
+func NewStackdriverConfigFromMap(config map[string]string) (*Config, error) {
+	sc := &Config{}
+
+	if pi, ok := config[ProjectIDKey]; ok {
+		sc.ProjectID = pi
+	}
+
+	if gl, ok := config[GcpLocationKey]; ok {
+		sc.GCPLocation = gl
+	}
+
+	if cn, ok := config[ClusterNameKey]; ok {
+		sc.ClusterName = cn
+	}
+
+	if gsn, ok := config[GcpSecretNameKey]; ok {
+		sc.GCPSecretName = gsn
+	}
+
+	if gsns, ok := config[GcpSecretNamespaceKey]; ok {
+		sc.GCPSecretNamespace = gsns
+	}
+
+	return sc, nil
+}
+
+// GetStackdriverSecret returns the Kubernetes Secret specified in the given config.
+func GetStackdriverSecret(config *Config) (*v1.Secret, error) {
+	ensureKubeclient()
+	ns := secretNamespaceDefault
+	if config.GCPSecretNamespace != "" {
+		ns = config.GCPSecretNamespace
+	}
+
+	return kubeclient.CoreV1().Secrets(ns).Get(config.GCPSecretName, metav1.GetOptions{})
+}
+
+// ConvertSecretToExporterOption converts a Kubernetes Secret to an OpenCensus Stackdriver Exporter Option.
+func ConvertSecretToExporterOption(secret *v1.Secret) option.ClientOption {
+	return option.WithCredentialsJSON(secret.Data[secretDataFieldKey])
+}
+
+// isKubeclientInitialized is a thread-safe check to see if the kubeclient is initialzed.
+// A reader-lock is used to make the check as efficient as possible.
+func isKubeclientInitialized() bool {
+	mtx.RLock()
+	defer mtx.RUnlock()
+
+	return kubeclient != nil
+}
+
+func ensureKubeclient() {
+	if isKubeclientInitialized() {
+		return
+	}
+
+	mtx.Lock()
+	defer mtx.Unlock()
+
+	// Multiple readers can read the kubeclient as nil before reaching
+	// the writer lock, so double check the value is not nil.
+	if kubeclient != nil {
+		return
+	}
+
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		panic(err.Error())
+	}
+
+	cs, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		panic(err.Error())
```

Comment:
Do we really want to panic?

Reply:
Probably? It really shouldn't happen on a Kubernetes cluster. The Kubernetes example panics https://github.com/kubernetes/client-go/blob/master/examples/in-cluster-client-configuration/main.go